python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py

```diff
@@ -239,19 +240,42 @@ def _task_definition(self, metadata, image):
             task_definition = self.ecs.describe_task_definition(taskDefinition=self.task_definition)
             return task_definition["taskDefinition"]
 
+        secrets = merge_dicts(
+            (
+                get_tagged_secrets(self.secrets_manager, self.secrets_tag)
+                if self.secrets_tag
+                else {}
+            ),
+            get_secrets_from_arns(self.secrets_manager, self.secrets),
+        )
+        secrets_dict = (
+            {"secrets": [{"name": key, "valueFrom": value} for key, value in secrets.items()]}
+            if secrets
+            else {}
+        )
+
+        task_definition = {}
+        with suppress(ClientError):
+            # TODO: Vary family name by repository location
+            task_definition = self.ecs.describe_task_definition(taskDefinition="dagster-run")[
+                "taskDefinition"
+            ]
+
+        container_definitions = task_definition.get("containerDefinitions", [{}])
+        for container_definition in container_definitions:
+            if (
+                container_definition.get("image") == image
+                and container_definition.get("name") == self.container_name
+                and container_definition.get("secrets") == secrets_dict.get("secrets")
+            ):
+                return task_definition
```

Comment:
One thing we lose by returning early is the ability to pick up cpu/memory changes to the daemon.

Previously, if you upsized your daemon, it'd automatically upsize your runs because each ensuing run would copy its cpu/memory from the daemon task definition.

Now, it'll be stuck with whatever the first definition's specs are until the image or the secrets changes. However, this remains overridable via tags. I think a nice improvement would be to more explicitly set defaults in the EcsRunLauncher's default task definition instead of borrowing cpu/memory from the daemon.

Reply:
To your point though - is there any way to make this use the full dictionary of config that uniquely defines the run, instead of just image and secrets?

I was imagining logic like:
- construct the dict that represents the task definition (default_ecs_task_definition minus the last line)
- construct that same dict for the existing task definition
- return existing if they're identical, create a new one if they are not

Reply:
The problem is that ECS DescribeTaskDefinitions returns a different dict shape than ECS RegisterTaskDefinition requires. The safest way to do this:

> construct the dict that represents the task definition (default_ecs_task_definition minus the last line)

Is to actually register the task definition. Otherwise, we need to maintain an understanding of ECS's internal logic for transforming our request into an actual task definition - something that feels like it'd be very brittle to API changes.

What we really need is for ECS to support find-or-create or create-if-not-exists or similar semantics. Short of that, I think we need to be picky about what we _actually_ care about for enforcing equality. I think image and secrets is a pretty good start but we can always add additional attributes to test for equality as we discover them.